lib/tail-file.js

```diff
@@ -0,0 +1,279 @@
+        , filename: this[kFileName]
+        , when: new Date()
+        })
+      } else if (this[kStartPos] !== eof) {
+        fileHasChanged = true
+      }
+
+      if (fileHasChanged) {
+        await this._streamFileChanges()
+      } else {
+        setImmediate(this.emit.bind(this), 'flush')
+      }
+
+      this._scheduleTimer(this[kPollFileIntervalMs])
+    } catch (err) {
+      if (err.code === 'ENOENT') {
+        if (this[kFileHandle]) {
+          // Don't lose the last lines in the file if it previously existed.
+          // Perhaps it has not been re-created yet (or won't be)
+          try {
+            await this._readRemainderFromFileHandle()
+          } catch (error) {
+            this.emit('tail_error', error)
+          }
+        }
+        this[kPollFailureCount]++
+        if (this[kPollFailureCount] >= this[kMaxPollFailures]) {
+          return this.quit(err)
+        }
+        this.emit('retry', {
+          message: 'File disappeared. Retrying.'
+        , filename: this[kFileName]
+        , attempts: this[kPollFailureCount]
+        , when: new Date()
+        })
+        this._scheduleTimer(this[kPollFailureRetryMs])
+        return
+      }
+      // Some other error like EACCES
+      // TODO: Retries for certain error codes can be put here
+      return this.quit(err)
+    }
+    return
+  }
+
+  _scheduleTimer(ms) {
+    clearTimeout(this[kPollTimer])
+    if (this[kQuitting]) return
+    this[kPollTimer] = setTimeout(this._pollFileForChanges.bind(this), ms)
+    return
+  }
+
+  async _streamFileChanges() {
+    try {
+      const stream = fs.createReadStream(this[kFileName], {
+        ...this[kOpts].readStreamOpts
+      , start: this[kStartPos]
+      })
+      await this._readChunks(stream)
+    } catch (err) {
+      // Possible file removal.  Let auto-retry handle it.
+      this[kPollFailureCount]++
+      const error = new Error('An error was encountered while tailing the file')
+      error.code = 'ETAIL'
+      error.meta = {actual: err}
+      // Emitting on 'error' would bork the parent Readstream
+      // FIXME: error can be emitted here
+      this.emit('tail_error', error)
+    }
+    return
+  }
+
+  _read() {
+    if (this[kStream]) {
+      this._readChunks(this[kStream])
+    }
+    return
+  }
+
+  async quit(err) {
+    this[kQuitting] = true
+    clearTimeout(this[kPollTimer])
+
+    if (err) {
+      this.emit('error', err)
+    } else {
+      // One last read to get lines added in high throughput
+      this._pollFileForChanges().catch(NOOP)
+      await once(this, 'flush')
+    }
+
+    // Signal the end of this Readstream
+    this.push(null)
+
+    // Clean open file handles and streams
+    if (this[kFileHandle]) {
+      this[kFileHandle].close().catch(NOOP)
+    }
+    if (this[kStream]) {
+      this[kStream].destroy()
+    }
+
+    process.nextTick(() => {
+      if (this._readableState && !this._readableState.endEmitted) {
```

Comment:
I do have some concerns about reaching into `_readableState`.... that being said, I don't believe there is a better way of doing this... :[

Reply:
Yeah, this took some tweaking to get it to work across versions.  I tried to use `readable.readableEnded`, but that's not available until node 12